src/common/tse.c

```diff
@@ -248,6 +249,11 @@ tse_task_decref(tse_task_t *task)
 
 	D_ASSERT(d_list_empty(&dtp->dtp_dep_list));
 
+	/* XXX: Do we need to trigger callback? */
+	while ((dtc = d_list_pop_entry(&dtp->dtp_comp_cb_list,
+				       struct tse_task_cb, dtc_list)) != NULL)
+		D_FREE(dtc);
```

Comment:
[Note] I don't know if this change is right or wrong; perhaps worth reviews from @mchaarawi and @wangdi1.

Reply:
In our current implementation, the task complete callback is triggered via tse_task_complete_callback(). But in some failure cases before schedule, some logic may call tse_task_decref() to directly release the "task" instead of via tse_task_complete(). Then former registered task complete callback handles will be leaked.

Here, I am not sure whether we need to trigger related callbacks or directly free DRAM for the callback handles. As my understand, if they want to trigger related complete callback, the logic needs to call tse_task_complete() instead of tse_task_decref(). So if they call tse_task_decref(), I prefer to directly release related callback handles.

Reply:
yea i think we should not use decref as the last call to free the task. this task call should have been private IMO and task ref count maintained by users if they want to. but since it's used now in DAOS stack outside of tse, could there be a possibility that we call decref in non fatal cases? IMO the completion cbs should  be called regardless of task completion result (success or fail), but im not too familiar of how decref has been used lately.

Reply:
>For some reason we failed at the 3rd step, but the logic does not want to trigger completion CB under such case, since we do not have the unregister API for completion CB.

Once the callback is registered to the task, then it has to be called in any failure cases, in case the callback is used to free some resource other than dtc.

Reply:
The registered CB may be used to release some resource, or may be used to trigger RPC resend. When we come here, we are releasing the last reference on the task, if related RPC resend is triggered, it may cause trouble, right? So I think we need to define some usage rules: it is the caller's duty to guarantee to call tse_task_complete() for triggering registered CB; the tse_task_decref() is only used to release reference on the task, if it is the last reference, then registered CB list must be empty. If yes, then we need to scan and check all tse_task_decref() usage against the rules.

Reply:
>The registered CB may be used to release some resource, or may be used to trigger RPC resend. When we come here, we are releasing the last reference on the task, if related RPC resend is triggered, it may cause trouble, right?

You mean resend of the task is triggered, but the task is freed, sounds like a "different" bug to me?  Sorry, not sure I follow here, could you please elaborate a bit?

> So I think we need to define some usage rules: it is the caller's duty to guarantee to call tse_task_complete() for triggering registered CB; 

Hmm, I tend to think this is the task/tse's rule, once the registered CB is registered, then it should be executed, even the task is never scheduled. 

>  the tse_task_decref() is only used to release reference on the task, if it is the last reference, then registered CB list must be empty. If yes, then we need to scan and check all tse_task_decref() usage against the rules.

Yes, agree.